### DISCUSSION TITLE
[已完成] [ 控件 ] up主 可不可以添加一个桌面歌词控件

### DISCUSSION THREAD

**Discussion by e0528** [GeekLee2012/Less-Player-Desktop#9] [Ideas]
up主 可不可以添加一个桌面歌词控件

**Comment by GeekLee2012** [comment]
**歌词逐字高亮**，效果做不出来啊。

普通的整行高亮，或根据歌词时长匀速高亮，这样的效果又不是很想做。

有点“ **高不成低不就** ”了，属于是......

**Reply by qianfeng0221** [reply]
其实个人感觉“歌词逐字高亮”功能也没必要做，逐句高亮就很好啦，如果是“根据歌词时长匀速高亮”的话，有其他播放器这么做过，但是效果不尽人意，最主要的原因还是现在国内音乐平台刮削下来的歌词格式的问题，都是一句一个时间轴，而不是一个字一个时间，所以也就没必要做“歌词逐字高亮“功能了

**Reply by GeekLee2012** [reply]
> 其实个人感觉“歌词逐字高亮”功能也没必要做，逐句高亮就很好啦，如果是“根据歌词时长匀速高亮”的话，有其他播放器这么做过，但是效果不尽人意，最主要的原因还是现在国内音乐平台刮削下来的歌词格式的问题，都是一句一个时间轴，而不是一个字一个时间，所以也就没必要做“歌词逐字高亮“功能了

“逐字歌词”，确实暂时没那么必要。
后续想做了，再去研究怎么做吧。 
主要目前逐字带时间的歌词，也没找到好的来源平台。

**Comment by GeekLee2012** [comment]
> **歌词逐字高亮**，效果做不出来啊。
> 
> 普通的整行高亮，或根据歌词时长匀速高亮，这样的效果又不是很想做。
> 
> 有点“ **高不成低不就** ”了，属于是......


有桌面歌词控件，确实体验会好些。
后续考虑好实现方式，再说吧。

**Comment by GeekLee2012** [comment]
朋友你好！“好久不见”！

v0.1.17版本已支持桌面歌词。该有的功能，虽迟必到！但不支持逐字歌词，也没有过多的动画效果。

而且，目前版本还有些平台兼容性问题，
比如**鼠标事件透传**问题，下面简称为**透传**。

桌面歌词在**锁定状态**下，
macOS平台下，完全支持透传。
而Windows平台下，虽然也支持，但**无法**在鼠标移动到**歌词上面**时，重新再显示“锁定/解锁”按钮。

所以，不同平台我采取了不同方式：
macOS平台完全支持透传，Windows等**非macOS平台不支持透传**。
那么，非macOS平台下，桌面歌词如果置顶显示，且处于锁定状态，
很可能会遮挡桌面歌词下面的其他软件窗口的操作，即使看起来透明的没有遮挡。

**简单来说，没有透传，那锁定状态下的桌面歌词窗口就像是一块透明的玻璃。**

**Comment by GeekLee2012** [comment]
桌面歌词窗口，支持少数应用内快捷键的：
左方向键 <-，播放上一曲  
右方向键 ->，播放下一曲  
空格键，播放 / 暂停  

显示模式为“全部”时，且在非锁定状态下，支持播放到指定歌词，
即鼠标滚动歌词时，当前滚动位置所在歌词行旁边会显示“歌词时间”以及播放按钮。  

<img width="441" alt="Screen Shot 2023-07-29 at 12 34 10" src="https://github.com/GeekLee2012/Less-Player-Desktop/assets/14144549/dd878b35-fe7f-420b-a61a-a7bd3dca5112">
